### PR TITLE
Various bug fixes and small enhancements

### PR DESCRIPTION
**Pull request by bmbroom**
Note: I have not addressed the keyboard navigation issues in this pull request.

Note: increment version number immediately prior to merging

### CODE REVIEW

**Review by jmelott — CHANGES_REQUESTED** [review]
I'd question the usefulness or desire to show the youtube videos in an iframe in the viewer.  That eliminates a lot of the usefulness of the videos of being able to play them alongside the ngchm while you're trying things out.  It stops access to the other elements on the youtube page like the time links and other things.  I think it would be better to open the playlist in another tab instead.

**Comment by bmbroom** [comment]
We can discuss Monday.  I find it's a more efficient use of screen space than having a separate window.  (At least on Mac and perhaps Windows. I don't know how to make a window "always on top" on either of those systems like you can on Linux.)

It's very easy to open a new tab from the popup by clicking on the "Watch on Youtube" button.  It's not possible to go the other way.

I don't understand your comment about not being able to watch the videos while trying things out.  Except for a couple of popups dismissing the videos, I was able to do so.  Those cases will have to be fixed if we decide to go ahead with this idea.

**Comment by jmelott** [comment]
I hadn't actually tried playing the video, but if it is not modal then people can still use the map, then that's OK.

**Review by jmelott — APPROVED** [review]

File: NGCHM/WebContent/javascript/Linkout.js [hunk 1]
```diff
@@ -816,10 +816,10 @@ var linkoutsVersion = 'undefined';
 			LNK.addLinkout("Copy selected labels to clipboard", rowLabelType[0], linkouts.MULTI_SELECT, LNK.copyToClipBoard,null,0);
 		}
 
-		LNK.addLinkout("Copy bar data for all labels", "ColumnCovar", null, LNK.copyEntireClassBarToClipBoard,null,0);
-		LNK.addLinkout("Copy bar data for selected labels", "ColumnCovar", linkouts.MULTI_SELECT,LNK.copyPartialClassBarToClipBoard,null,1);
-		LNK.addLinkout("Copy bar data for all labels", "RowCovar", null,LNK.copyEntireClassBarToClipBoard,null,0);
-		LNK.addLinkout("Copy bar data for selected labels", "RowCovar", linkouts.MULTI_SELECT,LNK.copyPartialClassBarToClipBoard,null,1);
+		LNK.addLinkout("Copy bar data for all columns", "ColumnCovar", linkouts.MULTI_SELECT, LNK.copyEntireClassBarToClipBoard,null,0);
```
Comment (ChrisWakefield) [inline NGCHM/WebContent/javascript/Linkout.js]: Change in 3rd arg from null to MULTI_SELECT is unclear.  Same issue just below.

File: NGCHM/WebContent/javascript/Linkout.js [hunk 2]
```diff
@@ -645,17 +645,21 @@ var linkoutsVersion = 'undefined';
 				LNK.addMenuItemToTable(axis, table, grpLinkouts[l].linkout, true);
 			}
 		} else {
-			//Always add clipboard link at top of list
-			LNK.addMenuItemToTable(axis, table, grpLinkouts[0].linkout, true);
+			const covar = axis.indexOf('Covar') != -1;
```
Comment (ChrisWakefield) [inline NGCHM/WebContent/javascript/Linkout.js:648]: i.e. axis is RowCovar or ColumnCovar.  A future rewrite of Linkout.js should eliminate magic name strings.

File: NGCHM/WebContent/javascript/Linkout.js [hunk 3]
```diff
@@ -660,6 +660,12 @@ var linkoutsVersion = 'undefined';
 			if (grpLinkouts.length > firstGroupLinkout) {
 				let addedHeader = false;
 				for (let l = firstGroupLinkout; l < grpLinkouts.length;l++ ) {
+					if (covar && grpLinkouts[l].linkout.title.indexOf('for selected') != -1 &&
```
Comment (ChrisWakefield) [inline NGCHM/WebContent/javascript/Linkout.js:663]: Another magic string.  Logic should be reversed (negated).

File: NGCHM/WebContent/javascript/DetailHeatMapEvent.js [hunk 4]
```diff
@@ -1002,7 +1002,18 @@ DEV.detailDataZoomIn = function (mapItem) {
...
 		const startIndex = Math.min(focusIndex,anchorIndex), endIndex = Math.max(focusIndex,anchorIndex);
-		SRCH.setAxisSearchResults (axis, startIndex, endIndex);
+		if (axis.indexOf('Covar') == -1) {
+		    SRCH.setAxisSearchResults (axis, startIndex, endIndex);
+		} else {
+		    // Can select covariate bar clicked on, but
+		    SRCH.setAxisSearchResults (axis, focusIndex, focusIndex);
+		    // Some intermediate covariate bars may be hidden, so need to test for presence of each one.
+		    for (let idx = startIndex+1; idx < endIndex; idx++) {
+			if (document.querySelector('.ClassBar[data-axis="' + axis + '"][data-index="' + idx + '"]')) {
```
Comment (ChrisWakefield) [inline NGCHM/WebContent/javascript/DetailHeatMapEvent.js:1012]: Using DOM state?  Is this the way we model shown/hidden?

File: NGCHM/WebContent/javascript/NGCHM_Util.js [hunk 5]
```diff
@@ -811,6 +811,9 @@ UTIL.getClickType = function (e) {
 	    case 3: clickType = 1;
 	    break; 
 	}
+	 if (e.ctrlKey) {
```
Comment (ChrisWakefield) [inline NGCHM/WebContent/javascript/NGCHM_Util.js:815]: No problem with this, but this function should be reviewed and rewritten or commented.

File: NGCHM/WebContent/javascript/Customization.js [hunk 6]
```diff
@@ -286,20 +286,34 @@
...
-	    var script = document.createElement('script');
+	CUST.beforeLoadCustom(heatMap);
+	const head = document.getElementsByTagName('head')[0];
+	const script = document.createElement('script');
+	if (CFG.custom_specified) {
+	    initScript();
+	} else {
+	    // USE DEFAULT CUSTOM.JS
+	    // NOTE: These 3 lines of code are replaced when building ngchmApp.html and ngchmWidget-min.js (the "file mode" and "widget" versions of the application).
+	    initScript();
```
Comment (ChrisWakefield) [inline NGCHM/WebContent/javascript/Customization.js:297]: i.e. the call to initScript() is edited out ?  (One?)

File: NGCHM/WebContent/javascript/Customization.js [hunk 7]
```diff
@@ -286,20 +286,34 @@
...
-        // Most browsers:   NOTE: The next 2 lines of code are replaced when building ngchmApp.html and ngchmWidget-min.js (the "file mode" and "widget" versions of the application)
-        script.onload = CUST.definePluginLinkouts;
-        // Internet explorer:
-        script.onreadystatechange = function() { if (this.readyState == 'complete') {     	CUST.definePluginLinkouts();   }   };  //Leave this as one line for filemode version app builder
-	script.onerror = function () {
-	    console.warn ("Loading of " + CFG.custom_script + " failed.");
-	    CUST.definePluginLinkouts();
-	};
-	head.appendChild(script);
+	    // Most browsers:
```
Comment (ChrisWakefield) [inline NGCHM/WebContent/javascript/Customization.js:304]: So the Minimizer no longer modifies this block?

File: NGCHM/WebContent/javascript/UI-Manager.js [hunk 8]
```diff
@@ -1273,6 +1272,7 @@
 			case 'ArrowLeft': // left key
 				if (document.activeElement.id !== "search_text"){
 					e.preventDefault();
+					if (!mapItem) return;
```
Comment (ChrisWakefield) [inline NGCHM/WebContent/javascript/UI-Manager.js:1305]: "mapItem" is always either =primaryMap or null ?  If that's true, I suggest naming the variable primaryMap.

File: NGCHM/WebContent/javascript/PdfGenerator.js [hunk 9]
```diff
@@ -543,13 +548,13 @@ PDF.setBuilderLogText = function (doc, text, pos, end) {
 		} else {
 			// If widgetized viewer exclude MDA logo and show compressed hear
```
Comment (ChrisWakefield) [inline NGCHM/WebContent/javascript/PdfGenerator.js:575]: Does this (old) comment make sense to anyone?  ("show compressed hear")

File: NGCHM/WebContent/javascript/PdfGenerator.js [hunk 10]
```diff
@@ -32,12 +32,22 @@ PDF.canGeneratePdf = function() {
 	return SUM.isVisible() || DVW.anyVisible();
 };
 
+PDF.pdfDialogClosed = function() {
```
Comment (ChrisWakefield) [inline NGCHM/WebContent/javascript/PdfGenerator.js:35]: Shouldn't "PDF Dialog" be modal?  With a function like jquery's modaldialog(), this logic isn't needed. (?)

File: NGCHM/WebContent/javascript/UserPreferenceManager.js [hunk 11]
```diff
@@ -354,19 +355,21 @@ UPM.doApply = function(isReset){
...
 	} else {
 		//When resetting no validations need be performed and, if they were, 
 		//additional modifications to validation logic would be required.
 		UPM.prefsApply();
 		heatMap.setUnAppliedChanges(true);
 		UPM.prefsSuccess();
-		UPM.enableApplyButton();
+		enableApplyButton();
 	}
 }
```
Comment (ChrisWakefield) [inline NGCHM/WebContent/javascript/UserPreferenceManager.js:374]: Factor out enableApplyButton() so it's clear it is called in all cases.

**Review by ChrisWakefield — APPROVED** [review]
Based on only code review, made a number of comments, but otherwise ok.

**Comment by bmbroom** [comment]
After discussion with Chris, I will merge this pull request now without incrementing the version number.  The version number will dump just before merging the third pull request in the current group.

**Pull request merged**